contracts/adapters/swap/astrovault/src/contract.rs

```diff
@@ -0,0 +1,552 @@
+
+    Ok(Asset::new(
+        deps.api,
+        &simulation_response.to.info.to_string(),
+        simulation_response.to.amount,
+    ))
+}
+
+// Queries the astrovault pool contracts to simulate a swap exact amount in with metadata
+fn query_simulate_swap_exact_asset_in_with_metadata(
+    deps: Deps,
+    asset_in: Asset,
+    swap_operations: Vec<SwapOperation>,
+    include_spot_price: bool,
+) -> ContractResult<SimulateSwapExactAssetInResponse> {
+    // Error if swap operations is empty
+    let Some(first_op) = swap_operations.first() else {
+        return Err(ContractError::SwapOperationsEmpty);
+    };
+
+    // Ensure asset_in's denom is the same as the first swap operation's denom in
+    if asset_in.denom() != first_op.denom_in {
+        return Err(ContractError::CoinInDenomMismatch);
+    }
+
+    let astrovault_router_contract_address = ASTROVAULT_ROUTER_ADDRESS.load(deps.storage)?;
+    let hops = convert_operations_to_hops(
+        deps,
+        astrovault_router_contract_address.to_string(),
+        swap_operations,
+    )?;
+
+    let simulation_response: QueryRouteSwapSimulation =
+        deps.querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: astrovault_router_contract_address.to_string(),
+            msg: to_json_binary(&router::query_msg::QueryMsg::RouteSwapSimulation {
+                amount: asset_in.amount(),
+                hops,
+            })?,
+        }))?;
+
+    // Create the response
+    let response = SimulateSwapExactAssetInResponse {
+        asset_out: Asset::new(
+            deps.api,
+            &simulation_response.to.info.to_string(),
+            simulation_response.to.amount,
+        ),
+        spot_price: if include_spot_price {
+            Some(simulation_response.to_spot_price)
+        } else {
+            None
+        },
+    };
+
+    Ok(response)
+}
+
+fn query_simulate_smart_swap_exact_asset_in(
+    deps: Deps,
+    ask_denom: String,
+    routes: Vec<Route>,
+) -> ContractResult<Asset> {
+    simulate_smart_swap_exact_asset_in(deps, ask_denom, routes)
+}
+
+fn simulate_smart_swap_exact_asset_in(
+    deps: Deps,
+    ask_denom: String,
+    routes: Vec<Route>,
+) -> ContractResult<Asset> {
+    let mut asset_out = Asset::new(deps.api, &ask_denom, Uint128::zero());
+
+    for route in &routes {
+        let route_asset_out = query_simulate_swap_exact_asset_in(
+            deps,
+            route.offer_asset.clone(),
+            route.operations.clone(),
+        )?;
+
+        asset_out.add(route_asset_out.amount())?;
+    }
+
+    Ok(asset_out)
+}
+
+fn query_simulate_smart_swap_exact_asset_in_with_metadata(
+    deps: Deps,
+    _asset_in: Asset,
+    ask_denom: String,
+    routes: Vec<Route>,
+    _include_spot_price: bool,
+) -> ContractResult<SimulateSmartSwapExactAssetInResponse> {
+    let asset_out = simulate_smart_swap_exact_asset_in(deps, ask_denom, routes.clone())?;
+
+    // TODO: Implement spot price calculation for smart swaps
```

Comment:
This still seems like an open TODO, should be able to do a weighted calculation of the spot prices to get the total route spot price output, lmk if need any support here